drivers/network/netio/netio.c

```diff
@@ -0,0 +1,903 @@
+        return NULL;
+    }
+    return ConnectionInformation;
+}
+
+static NTSTATUS WSKAPI
+WskBind(_In_ PWSK_SOCKET Socket, _In_ PSOCKADDR LocalAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    NTSTATUS status;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTRANSPORT_ADDRESS ta = TdiTransportAddressFromSocketAddress(LocalAddress);
+    if (ta == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    if (s->LocalAddressHandle != INVALID_HANDLE_VALUE)
+    {
+        ZwClose(s->LocalAddressHandle);
+        s->LocalAddressHandle = INVALID_HANDLE_VALUE;
+        s->LocalAddressFile = NULL;
+    }
+
+    status = TdiOpenAddressFile(&s->TdiName,
+                                ta, AFD_SHARE_REUSE, &s->LocalAddressHandle, &s->LocalAddressFile);
+    if (NT_SUCCESS(status))
+    {
+        memcpy(&s->LocalAddress, LocalAddress, sizeof(s->LocalAddress));
+    }
+    ExFreePoolWithTag(ta, 'ADDR');
+
+    Irp->IoStatus.Status = status;
+    return status;
+}
+
+enum direction
+{
+    DIR_SEND,
+    DIR_RECEIVE
+};
+
+static NTSTATUS WSKAPI
+WskSendTo(
+    _In_ PWSK_SOCKET Socket,
+    _In_ PWSK_BUF Buffer,
+    _Reserved_ ULONG Flags,
+    _In_opt_ PSOCKADDR RemoteAddress,
+    _In_ ULONG ControlInfoLength,
+    _In_reads_bytes_opt_(ControlInfoLength) PCMSGHDR ControlInfo,
+    _Inout_ PIRP Irp)
+{
+    PIRP tdiIrp = NULL;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo;
+    NTSTATUS status;
+    void *BufferData;
+    struct NetioContext *nc;
+
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    BufferData = MmGetSystemAddressForMdlSafe(Buffer->Mdl, NormalPagePriority);
+    if (BufferData == NULL)
+    {
+        DbgPrint("Error mapping MDL\n");
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
```

Comment:
The calling driver is neither required nor supposed to set `Thread`.
The fact that you're seeing a crash without this is a sign that somehow we're not calling the right completion routine.
IRP completion does the following:
* call the completion routine for each stack location
* if any completion routine returns `STATUS_MORE_PROCESSING_REQUIRED`, stop processing
* if there's another call to `IoCompleteRequest` after this, the remaining completion routines get called etc
* after the last stack location is processed in this way, assume the IRP is owned by the I/O manager and free the IRP -- this is where it will assert or BSOD if the I/O manager does not in fact own the IRP

So the BSOD indicates that the I/O manager tried to free an IRP that it did not own. An IRP allocated using `IoAllocateIrp` (which is usually where the IRP for WSK functions comes from) is owned by the allocating driver, not the I/O manager. That owning driver must therefore set a completion routine that returns `STATUS_MORE_PROCESSING_REQUIRED` so that the I/O manager will stop completion of the IRP at that point, so that it doesn't go to the I/O manager.

With this in mind, I'd suggest (re)reading https://learn.microsoft.com/en-us/windows-hardware/drivers/network/using-irps-with-winsock-kernel-functions and paying special attention to the `IoFreeIrp(Irp); return STATUS_MORE_PROCESSING_REQUIRED;` in the examples -- this is where this type of IRP is normally supposed to end, and I suspect you'll find that pattern in WinDRBD as well.

So what's likely happening here is one of two things:
* The calling driver forgot to set a completion routine (unlikely if that driver works on Windows)
* The current stack location is somehow off, so whenever you complete the IRP, the owning driver's completion routine gets skipped. This would indicate you're either doing too many or too few `IoCallDriver`/`IoSetNextIrpStackLocation` calls

Reply:
Hi: I get a 

    *** Fatal System Error: 0x0000007e

when the Thread pointer is NULL. The error happens at:

    ntoskrnl/ke/apc.c:676

where it does an ASSERT:

    675	        /* Sanity check */
    676	        ASSERT((TargetEnvironment <= Thread->ApcStateIndex) ||   
    677	               (TargetEnvironment == InsertApcEnvironment));

So my question would be: Is it legal for an Irp to have the Thread member uninitialized
(or simply NULL)? If not who is responsible for setting it? (WSK Application or NETIO?)